---
topic: functions
part: "5.1"
---
Author: [University of Helsinki](https://programming-23.mooc.fi/)
Type: Website
Topics: #python #mooc #functions 
Link: [Python Programming MOOC 2023](https://programming-23.mooc.fi/)
∗:[[00 PROGRAMMING/Python/_Python/Python]] 

---
# 01.5.5  Warning: Overwriting a parameter and returning too early

--- 
In programming, be cautious of overwriting parameters and returning too early from functions, as they can lead to unexpected bugs. Let's explore common pitfalls:

### Overwriting Parameters:

Example function that checks if an integer is in a list:

```python
def number_in_list(numbers: list, number: int):
    for number in numbers:
        if number == number:
            return True
        else:
            return False

```

- This function always returns ___True___ due to parameter overwriting.
- Rename the parameter to fix this issue.
```python
def number_in_list(numbers: list, searched_number: int):
    for number in numbers:
        if number == searched_number:
            return True
        else:
            return False
```
But there is a new problem, the function takes only the first item in the list,
and returns __True__ or __False__ depending on its value.
> We cannot know whether a number is not present in the list until we have checked all the items in the list.

The ___return False___ command should be placed outside the _for_ loop:
```python
def number_in_list(numbers: list, searched_number: int):
    for number in numbers:
        if number == searched_number:
            return True

    return False
```
### Early Returns:

Let's have a look at another faulty function:

```python
def unique_numbers(numbers: list):
    # a helper variable to store all the numbers we've already checked
    numbers = []
    for number in numbers:
        # have we seen this number already?
        if number in numbers:
            return False
        numbers.append(number)

    return True

unique = unique_numbers([1, 2, 2])
print(unique)  # prints out True
```
This function is supposed to check whether all numbers in a list are distinct from each other, but it always returns ___`True`___.

The function overwrites the value stored in its parameter by mistake.
The function tries to use the variable ___numbers___ to store all the numbers already checked, but this overwrites the original argument list.

Renaming the helper variable is an easy fix:
```python
def unique_numbers(numbers: list):
    # a helper variable to store all the numbers we've already checked
    numbers_checked = []
    for number in numbers:
        # have we seen this number already?
        if number in numbers_checked:
            return False
        numbers_checked.append(number)

    return True

unique = unique_numbers([1, 2, 2])
print(unique)  # prints out False
```


Example function to check if a number is unique in a list:

```python
def unique_numbers(numbers: list):
    # a helper variable to store checked numbers
    numbers = []
    for number in numbers:
        if number in numbers:
            return False
        numbers.append(number)
    return True

```

- This function always returns ___True___.
- It overwrites the argument list and checks uniqueness incorrectly.
- Rename the helper variable to fix this issue.